draft-wang-ppm-dap-taskprov.md

```diff
@@ -56,76 +60,330 @@ collector). Thus it is up to each deployment of DAP to securely implement its
+client.
+
+To offer the `task-prov` extension, the client adds the `TaskConfig` structure
+it received from the task author in the extensions field of its `Report`. It
+computes the task ID as described in {{construct-task-id}}.
+
+## Construct task ID {#construct-task-id}
+
+For `task-prov` extension, a DAP task is not created before distributing task
+configuration to clients. Therefore, a DAP task ID may not be available to
+clients, aggregators and collector before uploading. When the `task-prov`
+extension is used, the task ID is computed as follows:
+
+~~~
+task_id = SHA-256(task_config)
+~~~
+
+Task ID can be constructed on server side and then distributed to clients,
+or constructed independently on clients and servers. When constructed
+independently, the mechanism used for creating the task ID must be known to
+both clients and the collector.
 
 # Leader Behavior
 
 > TODO(wangshan) Say what the Leader does when it receives a report with this
 > extension. What do Leaders do if they *don't* support this extension? Is it
 > safe to ignore it? How are the upload, aggregate, and collect flows impacted?
 
+Leader should have saved any information that is not always tied to a
+particular task, for e.g. `vdaf_verify_key`. Leader may not know the task ID of
+the current task before receiving the first upload request.
+
+## Change to update sub-protocol
+
+Upon receiving a report, leader reads the extension codepoint in
+`extension_type`. If leader does not support this extension, it SHOULD abort
+the upload protocol and alert the client with error "unrecognizedMessage".
+
+> reason for aborting: if leader ignores it, then there is no way to ensure
+> clients that the extension is used and task parameters are respected in
+> aggregators
+
+If leader supports `task-prov` extension, it should first check if the task ID
+already exists, if so leader continues to the rest of upload flow as usual.
+Note if the existing tasks's configuration is different from the one in report
+extension, HPKE decryption will fail due to mismatched AAD.
+
+If the task ID has not been seen before, leader should read and decode report's
+`extension_data` with the `TaskConfig` schema. If the decoding failed, it MUST
+abort the upload protocol and alert the client with error
+"unrecognizedMessage".
+
+If the decode succeeds, leader should create a new task using the task ID from
+the decoded report, and save task configuration with the newly created task. In
+particular, leader should deserialize `vdaf_config` corresponding to `vdaf_type`,
+and pass the relevant parameters to the VDAF initializer. At this point, the
+task provision step has completed, and leader should continue to the rest of
+upload flow.
+
+Leader MAY return error to the client if task creation failed.
+
+> OPEN ISSUE: would returning error reveal the position of the client in the
+> batch of reports received? if so is this a security or privacy threat?
```

Comment:
Can you elaborate more on "position of the client in the batch of reports"? Is there anything that the client learns from the error response that it couldn't compute on its own?

Reply:
I can't remember what I was worried about, let me delete this and add back later with example